library/src/armory/metrics/detection.py

```diff
@@ -0,0 +1,189 @@
+"""Object detection metrics"""
+
+from typing import List, Optional, Sequence
+
+import torch
+from torchmetrics import Metric
+from torchmetrics.functional.detection import intersection_over_union
+from torchmetrics.utilities import dim_zero_cat
+
+from armory.data import BBoxFormat, BoundingBoxes
+from armory.metric import PredictionMetric
+
+
+class ObjectDetectionRates(Metric):
+    """
+    Metric for the following object detection metrics:
+
+    - True positive rate: the percent of ground-truth boxes which are predicted
+      with iou > iou_threshold, score > score_threshold, and the correct label
+    - Misclassification rate: the percent of ground-truth boxes with are
+      predicted with iou > iou_threshold, score > score_threshold, and the
+      incorrect label
+    - Disappearance rate: 1 - true positive rate - misclassification rate
+    - Hallucinations: the number of predicted boxes per image that have
+      score > score_threshold and iou < iou_threshold for each ground-truth box
+    """
+
+    @classmethod
+    def create(cls, *args, **kwargs):
+        """
+        Creates an instance of the object detection rates metric pre-wrapped in
+        a `armory.metric.PredictionMetric`
+        """
+        return PredictionMetric(
+            cls(*args, **kwargs), BoundingBoxes.as_torch(format=BBoxFormat.XYXY)
+        )
+
+    def __init__(
+        self,
+        mean: bool = True,
+        iou_threshold: float = 0.5,
+        score_threshold: float = 0.5,
+        class_list: Optional[Sequence[int]] = None,
+    ):
+        """
+        Initialize the metric.
+
+        Args:
+            mean: `True` to return the mean values across all images or `False`
+                to return the individual values per image
```

Comment:
MLFlow has a limitation in that _all_ metrics have to be scalar values. You _could_ treat the array indices as step values for the MLFlow metric--but that's just abusing the API I think (it has been recognized as a potential workaround, but not applicable in all cases: https://github.com/mlflow/mlflow/issues/2451).

We had a discussion about this many months ago during JATIC and we had said then that we were ok with metrics being wonky as presented in MLFlow because we didn't expect anyone to be trying to use them within MLFlow itself. If that assumption is being changed, then we certainly can revisit that (which has implications on all metrics from torchmetrics, not just this one). It strikes me as odd (i.e., variable behavior) if some metrics are directly recorded in MLFlow (i.e., when scalar) but others are simply attached as artifacts (i.e., when dicts & lists), but that's not a showstopper.